### PR TITLE
Task#6 Maybe

### PR DESCRIPTION
**Pull request by KroosT**
Created class _Maybe_ which may contain or not contain value.
It contain _isDefined()_ method that returns _true_ if value is present and _getOrThrow()_ method that throws exception if _isDefined() == false_ and returns value otherwise.

### CODE REVIEW

File: Tasks/src/Maybe/Maybe.java [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+    public static <T> Maybe<T> just(final T value) {
+        Validator.isArgNotNull(value, "value");
+        return new Maybe<>(value);
+    }
+
+    public static <T> Maybe<T> empty() {
+        return new Maybe<>((T) null);
+    }
+
+    public T getValueOrThrow() {
```
Comment (vmatelsky) [inline Tasks/src/Maybe/Maybe.java]: what about `getOrThrow`?

File: Tasks/src/Maybe/MaybeTest.java [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+
+import java.util.NoSuchElementException;
+
+public class MaybeTest {
+
+    private final Maybe<Integer> mIntegerMaybe = Maybe.just(229);
+    private final Maybe<Integer> mNullMaybe = Maybe.empty();
+
+    @Test
+    public void testGet() throws Exception {
```
Comment (vmatelsky) [inline Tasks/src/Maybe/MaybeTest.java]: split into two tests with meaningfull names

File: Tasks/src/Maybe/MaybeTest.java [hunk 3]
```diff
@@ -0,0 +1,40 @@
+package Maybe;
+
+import org.junit.Assert;
+import org.junit.Test;
+
+import java.util.NoSuchElementException;
+
+public class MaybeTest {
+
+    private final Maybe<Integer> mIntegerMaybe = Maybe.just(229);
```
Comment (vmatelsky) [inline Tasks/src/Maybe/MaybeTest.java]: move these inside tests

File: Tasks/src/Maybe/MaybeTest.java [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+    private final Maybe<Integer> mNullMaybe = Maybe.empty();
+
+    @Test
+    public void testGetOfIntegerMaybe() throws Exception {
+        Integer integerValue = null;
+
+        if (mIntegerMaybe.isDefined()) {
```
Comment (vmatelsky) [inline Tasks/src/Maybe/MaybeTest.java]: `isDefined` should be tested in two separate tests

File: Tasks/src/Maybe/MaybeTest.java [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+            integerValue = mIntegerMaybe.getOrThrow();
```
Comment (vmatelsky) [inline Tasks/src/Maybe/MaybeTest.java]: same for `getOrThrow`

File: Tasks/src/Maybe/MaybeTest.java [hunk 6]
```diff
@@ -1,40 +1,35 @@
...
-        Integer integerValue = null;
-
-        if (mIntegerMaybe.isDefined()) {
-            integerValue = mIntegerMaybe.getOrThrow();
-        }
-
-        Assert.assertEquals(integerValue, Integer.valueOf(229));
+    public void testIfDefinedReturnsTrue() {
+        final Maybe<Integer> integerMaybe = Maybe.just(229);
+        assertEquals(integerMaybe.isDefined(), true);
```
Comment (vmatelsky) [inline Tasks/src/Maybe/MaybeTest.java]: `assertTrue` ?

File: Tasks/src/Maybe/MaybeTest.java [hunk 7]
```diff
@@ -1,40 +1,35 @@
...
     @Test
-    public void testGetOfIntegerMaybe() throws Exception {
-        Integer integerValue = null;
-
-        if (mIntegerMaybe.isDefined()) {
-            integerValue = mIntegerMaybe.getOrThrow();
-        }
-
-        Assert.assertEquals(integerValue, Integer.valueOf(229));
+    public void testIfDefinedReturnsTrue() {
```
Comment (vmatelsky) [inline Tasks/src/Maybe/MaybeTest.java]: naming

File: Tasks/src/Maybe/MaybeTest.java [hunk 8]
```diff
@@ -1,40 +1,35 @@
...
     @Test
-    public void testGetOfNullMaybe() throws Exception {
-        Integer nullValue = null;
-
-        if (mNullMaybe.isDefined()) {
-            nullValue = mNullMaybe.getOrThrow();
-        }
+    public void testIfDefinedReturnsFalse() {
+        final Maybe<Integer> emptyMaybe = Maybe.empty();
+        assertEquals(emptyMaybe.isDefined(), false);
```
Comment (vmatelsky) [inline Tasks/src/Maybe/MaybeTest.java]: `assertElse`

File: Tasks/src/Maybe/MaybeTest.java [hunk 9]
```diff
@@ -1,40 +1,35 @@
...
+    }
 
-        Assert.assertEquals(nullValue, null);
+    @Test
+    public void testGetOrThrowReturnsValue() throws Exception {
+        final Maybe<Integer> integerMaybe = Maybe.just(229);
+        assertEquals(integerMaybe.getOrThrow(), Integer.valueOf(229));
```
Comment (vmatelsky) [inline Tasks/src/Maybe/MaybeTest.java]: `Integer.valueOf` why do we need it here?

Comment (KroosT) [reply Tasks/src/Maybe/MaybeTest.java]: Because _getOrThrow()_ returns Integer value.

File: Tasks/src/Maybe/Maybe.java [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+        mValue = value;
+    }
+
+    public static <T> Maybe<T> just(final T value) {
+        Validator.isArgNotNull(value, "value");
+        return new Maybe<>(value);
+    }
+
+    public static <T> Maybe<T> empty() {
+        return new Maybe<>((T) null);
```
Comment (vmatelsky) [inline Tasks/src/Maybe/Maybe.java:29]: could we avoid creating new instance every time this method called?

**Comment by vmatelsky** [comment]
LGTM

**Pull request merged**